analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs

```diff
@@ -125,25 +125,18 @@ protected override ProgramState PreProcessSimple(SymbolicContext context)
         {
             return context.State.SetOperationConstraint(operation, CollectionConstraint.Empty);
         }
-        else if (operation.AsInvocation() is { Instance: not null } invocation)
+        else if (operation.AsInvocation() is { } invocation)
         {
-            if (RaisingMethods.Contains(invocation.TargetMethod.Name))
-            {
-                if (context.State[invocation.Instance]?.HasConstraint(CollectionConstraint.Empty) is true)
-                {
-                    emptyAccess.Add(operation);
-                }
-                else
-                {
-                    nonEmptyAccess.Add(operation);
-                }
-            }
-            if (AddMethods.Contains(invocation.TargetMethod.Name) && invocation.Instance.TrackedSymbol() is { } symbol)
-            {
-                return context.State.SetSymbolConstraint(symbol, CollectionConstraint.NotEmpty);
-            }
+            return ProcessMethod(context, invocation.TargetMethod, invocation.Instance);
+        }
+        else if (operation.AsMethodReference() is { } methodReference)
+        {
+            return ProcessMethod(context, methodReference.Method, methodReference.Instance);
+        }
```

Comment:
I don't think we should treat `DoSomething(list.Add)` the same way as `list.Add()`. The former should remove `CollectionSonstraint`, and the latter should set `NotEmpty`.

Reply:
Why? What's the practical difference?

If someone sends `list.Add` somewhere, we should assume NotEmpty.

Reply:
Technically we don't know, whether or not `DoSomething` calls `list.Add`. But since we do treat `NotEmpty` and "no constraint" exactly the same, it doesn't matter.

Reply:
Then again... for `DoSomething(list.Clear())` it makes a bigger difference. Consider the following FP:
```
CallIfTrue(list.Clear, condition);
void CallIfTrue(Action action, bool condition)
{ if (condition) action(); }
```

I guess this kind of error is acceptable to keep the code simple, but let's add a few test cases for it.